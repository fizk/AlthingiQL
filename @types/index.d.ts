
export interface Assembly {
    id: number;
    period?: Period;
    division?: Division;
    cabinet?: Cabinet;
}

export interface Period {
    from: string;
    to: string;
}

export interface Division {
    majority: Party[];
    minority: Party[];
}

export interface Cabinet {
    id: number;
    title: string;
    description: string;
    period: Period;
    assemblies: Assembly[];
}

export interface Party {
    id: number;
    name: string;
    abbr_short?: string; //@todo fix naming
    abbr_long?: string; //@todo fix naming
    color?: string;
}

export interface PartyTime {
    party: Party;
    time: number;
}

export interface Category {
    id: number,
    description: string,
    superId: number,
    title: string;
}

export interface CategoryCount {
    count: number;
    category: Category;
}

export interface SuperCategory {
    id: number,
    title: string;
}

export interface IssueType {
    id: number;
    assembly: Assembly;
    name: string;
    type: Type
    date: string;
    speakers: CongressmanValue[];
    // speechRange: SpeechRange[];
    speechCount: number | null;
    speechTime: number | null;
}

export interface IssueCount {
    order: number,
    type: string,
    typeName: string,
    typeSubName: string,
    documentType: string,
    count: number,
}

export interface IssueA extends IssueType {
    subName: string;
    status: string;
    question: string;
    goal: string;
    majorChanges: string;
    changesInLaw: string;
    costsAndRevenues: string;
    deliveries: string;
    additionalInformation: string;
    proponents: Proponent[];
    proponentsCount: number;
    // voteRange: VoteRange[];
    categories: Category[],
    superCategories: SuperCategory[],
    governmentIssue: boolean,
}

export interface IssueB extends IssueType {

}

export interface Type {
    type:  string;//'n' | 'b' | 'l' | 'm' | 'q' | 's' | 'v' | 'a' | 'f' | /**/ 'ff' | 'ft' | 'um' | 'ud' | 'uu';
    category: string;//'a' | 'b'
    typeName: string
    typeSubName: string
}

export interface TypeCount {
    count: number;
    type: Type;
}

export type Issue = IssueA | IssueB;

export interface IssueValue {
    value: number,
    issue: Issue
}

export interface CongressmanValue {
    congressman: Congressman;
    value: number;
}

export interface Speech {
    id: string;
    plenaryId: number;
    assembly: Assembly;
    issue: Issue;
    congressman: Congressman;
    congressmanType: string;
    period: Period;
    text: string;
    type: string;
    iteration: string;
    position: number;
}

export interface SpeechRange {
    count: number;
    date: string;
}

export interface Person {
    id: number;
    name: string;
    avatar: Picture;
}

export interface Congressman extends Person {
    party: Party;
    assembly?: Assembly;
    constituency?: Constituency
}

export interface Proponent extends Congressman {
    order: number;
    minister: string | null
}

export interface Vote {
    id: number;
    voteId: number;
    vote: string;
    congressman: Congressman;
}

export interface VoteRange {
    count: number;
    date: string;
}

export interface Picture {
    src?: string;
    templateSrc: string;
}

export interface Plenary {
    id: number;
    assembly: Assembly;
    from: string;
    to: string;
    name: string;
    agenda: PlenaryItem[];
}

export interface PlenaryItem {
    id: number;
    plenaryId: number;
    assembly: Assembly;
    category: string;
    iterationType: string;
    iterationContinue: string;
    iterationComment: string;
    comment: string;
    commentType: string;
    posedId: number;
    answerer: string;
    issue: Issue;
    posedCongressman: Congressman;
    answererCongressman: Congressman;
    counterAnswererCongressman: Congressman;
    instigatorCongressman: Congressman;
}

export interface Inflation {
    id: number;
    value: number;
    date: string;
}

export interface Session {
    id: number;
    congressman: Congressman;
    assembly: Assembly;
    constituency: Constituency;
    party: Party;
    period: Period;
    type: string;
    abbr: string;
}

export interface Constituency {
    id: number;
    name: string;
    abbr_short: string; //todo fix naming
    abbr_long: string; //todo fix naming
    description: string;
}

export interface CategorySummary {
    id: number;
    superCategoryId: number;
    title: string;
    count: number;
}

export interface CategorySpeechTime {
    category: {
        id: number,
    },
    superCategory: {
        id: number,
    },
    title: string,
    time: number,
}

export interface SuperCategorySpeechTime {
    superCategory: {
        id: number,
        title: string,
    },
    time: number,
}

export interface IssueSummary {
    count: number;
    type: string;
    typeName: string;
    typeSubName: string;
}

export interface VoteSummary {
    count: number;
    vote: string;
}

export interface Document {
    id: number;
    issueId: number;
    assemblyId: number;
    date: string;
    url: string;
    type: string;
    proponents: Congressman[];
    votes: VoteResult[];
}

export interface VoteResult {
    id: number;
    issue_id: number;
    assembly_id: number;
    document_id: number;
    date: string;
    type: string;
    outcome: string;
    method: string;
    yes: number;
    no: number;
    inaction: number;
    committee: string;
}

export interface StatusCount {
    count: number,
    status: string
}

export interface ServerFetchStatus {
    error: any;
    loading: boolean;
}

export interface Progress {
    issue: {
        id: number;
        category: string;
    };
    assembly: {
        id: number;
    };
    committee: {
        id: number;
        name: string;
    };
    speech: {
        id: string;
    };
    document: {
        id: number;
    };
    date: string;
    title: string;
    type: string;
    completed: boolean;
}

export interface ClientCursor {
    to: number;
    from: number;
}

export interface Config {
    host: string;
    port: string | number;
}

export interface Client {
    getPagination: (url: string, cursor: any) => Promise<{data: object, cursor: {from: number, to: number}, done: boolean}>
    get: (url: any) => Promise<object>
}

export declare namespace DataSource {
    export interface Party {
        party_id: number;
        name: string;
        abbr_short: string;
        abbr_long: string;
        color: string;
    }

    export interface Cabinet {
        cabinet_id: number;
        title: string;
        description: string;
        from: string;
        to: string;
        assemblies?: Assembly[]
    }

    export interface Assembly {
        assembly_id: number;
        from: string;
        to: string;
        party: {
            majority: Party[];
            minority: Party[];
        };
        cabinet: Cabinet
    }
    
    export interface Committee {
        committee_id: number;
        name: string;
        first_assembly_id: number;
        last_assembly_id: number;
        abbr_long: string;
        abbr_short: string;
    }
    
    export interface Constituency {
        abbr_long: string;
        abbr_short: string;
        constituency_id: number;
        description: string;
        name: string;
    }
    
    export interface Issue {
        issue_id: number;
        assembly_id: number;
        congressman_id: number;
        category: 'A' | 'B';
        name: string;
        sub_name: string;
        type: string;
        type_name: string;
        type_subname: string;
        status: string;
        question: string;
        goal: string;
        major_changes: string;
        changes_in_law: string;
        costs_and_revenues: string;
        deliveries: string;
        additional_information: string;
        date: string;
        proponents: Proponent[];
        // voteRange: DateAndCount[];
        // speechRange: DateAndCount[];
        speakers: CongressmanValue[]
        categories: Category;
        super_categories: SuperCategory[]
        government_issue: boolean,
        speech_count: number,
        speech_time: number,
        issue_links: {
            assembly_id: number;
            issue_id: number;
            category: 'A' | 'B';
            type: string;
        }[]
    }
    
    export interface DateAndCount {
        date: string;
        count: number;
    }

    export interface Person {
        congressman_id: number;
        name: string;
        birth: string;
        death: string;
        abbreviation: string;
    }
    
    export interface Congressman extends Person {
        constituency: Constituency;
        party: Party;
    }

    export interface Proponent extends Congressman {
        minister: string | null;
        order: number;
    }

    export interface CongressmanValue extends Congressman {
        value: number;
    }

    export interface CongressmanAndDateRange extends Congressman{
        time: number;
        begin: string;
        end: string;
    }
    
    export interface Category {
        category_id: number;
        super_category_id: number;
        title: string;
        description: string;
    }

    export interface SuperCategory {
        super_category_id: number;
        title: string;
    }
}

declare global {
    interface Window {
        __APOLLO_STATE__: any;

    }
}
