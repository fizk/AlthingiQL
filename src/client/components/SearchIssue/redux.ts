export default () => {}
// const ISSUE_SEARCH = {
//     BEGIN: 'ISSUE_SEARCH:BEGIN',
//     END: 'ISSUE_SEARCH:END',
//     ERROR: 'ISSUE_SEARCH:ERROR',
//     CLEAR: 'ISSUE_SEARCH:CLEAR',
// };
//
// const issueSearchActionBegin = () => ({
//     type: ISSUE_SEARCH.BEGIN,
// });
//
// const issueSearchActionEnd = result => ({
//     type: ISSUE_SEARCH.END,
//     result,
// });
//
// const issueSearchActionError = error => ({
//     type: ISSUE_SEARCH.ERROR,
//     error,
// });
//
// export const issueClearAction = () => ({
//     type: ISSUE_SEARCH.CLEAR,
// });
//
// const searchQuery = `
//     query searchAssemblyIssue ($assembly: Int! $query: String!) {
//         SearchAssemblyIssue (assembly: $assembly, query: $query) {
//             id
//             assembly {id}
//             category
//             name
//             subName
//             type
//             typeName
//             typeSubName
//             status
//             question
//             goal
//             majorChanges
//             changesInLaw
//             costsAndRevenues
//             deliveries
//             additionalInformation
//             date
//             proponents {
//                 id
//                 name
//                 party {
//                     id
//                     name
//                     color
//                 }
//             }
//         }
//     }
// `;
//
// export const issueSearchAction = () => {
//     const threshold = 1000;
//     let last;
//     let deferTimer;
//     return (assembly, issue, query) => dispatch => {
//         const now = + new Date();
//         if (last && now < last + threshold) {
//             clearTimeout(deferTimer);
//             deferTimer = setTimeout(() => {
//                 last = now;
//                 issueSearchDispatchAction(dispatch, assembly, issue, query);
//             }, threshold);
//         } else {
//             last = now;
//             issueSearchDispatchAction(dispatch, assembly, issue, query);
//         }
//     };
// };
//
// const issueSearchDispatchAction = (dispatch, assembly, issue, query) => {
//     dispatch(issueSearchActionBegin());
//
//     const headers = new Headers();
//     headers.append('Content-Type', 'application/json');
//
//     const req = new Request('http://localhost:3000/graphql', {
//         method: 'POST',
//         body: JSON.stringify({
//             query: searchQuery,
//             variables: {assembly, issue, query},
//         }),
//         headers,
//     });
//
//     return fetch(req)
//         .then(response => response.json())
//         .then(json => {
//             if (json.hasOwnProperty('errors')) {
//                 dispatch(issueSearchActionError(json.errors));
//             } else {
//                 dispatch(issueSearchActionEnd(json.data.SearchAssemblyIssue));
//             }
//         })
//         .catch(error => dispatch(issueSearchActionError(error)));
// };
//
// export default (state = {result: [], loading: false, error: false}, action) => {
//     switch (action.type) {
//         case ISSUE_SEARCH.BEGIN:
//             return {
//                 ...state,
//                 loading: true,
//                 error: false,
//             };
//         case ISSUE_SEARCH.END:
//             return {
//                 result: action.result,
//                 loading: false,
//                 error: false,
//             };
//         case ISSUE_SEARCH.ERROR:
//             return {
//                 ...state,
//                 loading: false,
//                 error: action.error,
//             };
//         case ISSUE_SEARCH.CLEAR:
//             return {
//                 result: [],
//                 loading: false,
//                 error: false,
//             };
//         default:
//             return state;
//     }
// };
