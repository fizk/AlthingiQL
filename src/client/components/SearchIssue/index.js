import React from 'react';
import PropTypes from 'prop-types';
import {Redirect} from 'react-router-dom';
import {OptionsWithKeyBinding} from '../../elements/Form';
import {connect} from 'react-redux';
import {issueSearchAction, issueClearAction} from './redux';
import {IssueSearchResult} from '../../elements/SearchResult';

class SearchIssue extends React.Component {
    static propTypes = {
        assembly: PropTypes.number,
        result: PropTypes.arrayOf(PropTypes.shape({
            id: PropTypes.number,
            assembly: PropTypes.shape({id: PropTypes.number}),
            category: PropTypes.string,
            name: PropTypes.string,
            subName: PropTypes.string,
            type: PropTypes.string,
            typeName: PropTypes.string,
            typeSubName: PropTypes.string,
            status: PropTypes.string,
            question: PropTypes.string,
            goal: PropTypes.string,
            majorChanges: PropTypes.string,
            changesInLaw: PropTypes.string,
            costsAndRevenues: PropTypes.string,
            deliveries: PropTypes.string,
            additionalInformation: PropTypes.string,
            date: PropTypes.string,
            proponents: PropTypes.arrayOf(PropTypes.shape({
                id: PropTypes.number,
                name: PropTypes.string,
                party: PropTypes.shape({
                    id: PropTypes.number,
                    name: PropTypes.string,
                    color: PropTypes.string,
                }),
            }))
        })),
        isSearching: PropTypes.bool,
        onSearch: PropTypes.func,
        onClear: PropTypes.func,
    };

    static defaultProps = {
        assembly: undefined,
        issue: undefined,
        result: [],
        isSearching: false,
        onSearch: () => {},
        onClear: () => {},
    };

    constructor(props) {
        super(props);

        this.handleOnChange = this.handleOnChange.bind(this);
        this.handleOnSelect = this.handleOnSelect.bind(this);
        this.handleOnClear = this.handleOnClear.bind(this);
        this.state = {
            redirect: undefined
        }
    }

    handleOnChange(value) {
        if (value.trim() !== '') {
            this.props.onSearch(value);
        }
    }

    handleOnSelect(value) {
        this.props.onClear();
        this.setState({redirect: value});
    }

    handleOnClear() {
        this.props.onClear()
    }

    render() {
        return (
            <div>
                {this.state.redirect && <Redirect push={true} to={`/loggjafathing/${this.state.redirect.assembly.id}/thingmal/${this.state.redirect.id}`} />}
                <OptionsWithKeyBinding
                    isSearching={this.props.isSearching}
                    onClear={this.handleOnClear}
                    onChange={this.handleOnChange}
                    onSelect={this.handleOnSelect}>
                    {this.props.result.map(result => (
                        <IssueSearchResult key={`search-result-${result.id}`} issue={result} />
                    ))}
                </OptionsWithKeyBinding>
            </div>
        )
    }
}

const mapStateToProps = state => {
    return {
        result: state.issueSearch.result,
        isSearching: state.issueSearch.loading,
    }
};

const mapDispatchToProps = () => {
    const throttleIssueSearchAction = issueSearchAction();
    return (dispatch, ownProps) => ({
        onSearch: query => dispatch(throttleIssueSearchAction(ownProps.assembly, ownProps.issue,  query)),
        onClear: () => dispatch(issueClearAction())
    })
};

const SearchIssueWithStore = connect(mapStateToProps, mapDispatchToProps())(SearchIssue);

export {SearchIssue, SearchIssueWithStore}
