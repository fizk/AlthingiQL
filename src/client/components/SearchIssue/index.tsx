import * as React from "react";
import { Fragment } from "react";
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import { OptionsWithKeyBinding } from "../../elements/Form";
import { issueSearchAction, issueClearAction } from "./redux";
import { IssueSearchResult } from "../../elements/SearchResult";

type SearchIssueProps = {
    assembly?: number,
    result?: {
        id?: number,
        assembly?: {
            id?: number
        },
        category?: string,
        name?: string,
        subName?: string,
        type?: string,
        typeName?: string,
        typeSubName?: string,
        status?: string,
        question?: string,
        goal?: string,
        majorChanges?: string,
        changesInLaw?: string,
        costsAndRevenues?: string,
        deliveries?: string,
        additionalInformation?: string,
        date?: string,
        proponents?: {
            id?: number,
            name?: string,
            party?: {
                id?: number,
                name?: string,
                color?: string
            }
        }[]
    }[],
    isSearching?: boolean,
    onSearch?: (...args: any[]) => any,
    onClear?: (...args: any[]) => any
};

type SearchIssueState = {
    redirect: any | undefined
};

export default class SearchIssue extends React.Component<SearchIssueProps, SearchIssueState> {
    static defaultProps = {
        assembly: undefined,
        issue: undefined,
        result: [],
        isSearching: false,
        onSearch: () => {},
        onClear: () => {}
    };
    constructor(props) {
        super(props);
        this.handleOnChange = this.handleOnChange.bind(this);
        this.handleOnSelect = this.handleOnSelect.bind(this);
        this.handleOnClear = this.handleOnClear.bind(this);
        this.state = {
            redirect: undefined
        };
    }
    handleOnChange(value) {
        if (value.trim() !== "") {
            this.props.onSearch(value);
        }
    }
    handleOnSelect(value) {
        this.props.onClear();
        this.setState({ redirect: value });
    }
    handleOnClear() {
        this.props.onClear();
    }
    render() {
        return (
            <Fragment>
                {this.state.redirect && (
                    <Redirect
                        push={true}
                        to={`/loggjafarthing/${
                            this.state.redirect.assembly.id
                        }/thingmal/${this.state.redirect.id}`}
                    />
                )}
                <OptionsWithKeyBinding
                    isError={false}
                    isSearching={this.props.isSearching}
                    onClear={this.handleOnClear}
                    onChange={this.handleOnChange}
                    onSelect={this.handleOnSelect}
                >
                    {this.props.result.map(result => (
                        <IssueSearchResult
                            key={`search-result-${result.id}`}
                            issue={result}
                        />
                    ))}
                </OptionsWithKeyBinding>
            </Fragment>
        );
    }
}
const mapStateToProps = state => {
    return {
        result: state.issueSearch.result,
        isSearching: state.issueSearch.loading
    };
};
const mapDispatchToProps = () => {
    const throttleIssueSearchAction = issueSearchAction();
    return (dispatch, ownProps) => ({
        onSearch: query =>
            dispatch(
                throttleIssueSearchAction(
                    ownProps.assembly,
                    ownProps.issue,
                    query
                )
            ),
        onClear: () => dispatch(issueClearAction())
    });
};
const SearchIssueWithStore = connect(
    mapStateToProps,
    mapDispatchToProps()
)(SearchIssue);
export { SearchIssueWithStore };
