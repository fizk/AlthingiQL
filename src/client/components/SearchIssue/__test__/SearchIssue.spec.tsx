import * as React from 'react';
import SearchIssue from '../index';
import renderer from 'react-test-renderer';
import {BrowserRouter as Router} from 'react-router-dom';
import {Issue} from '../../../../../@types';

type Partial<T> = { [P in keyof T]?: T[P]; };
type PartialIssue = Partial<Issue>;

describe('<SearchIssue />', () => {
    test('default props', () => {
        const tree = renderer.create(
            <Router >
                <SearchIssue assembly={1} />
            </Router>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    test('data', () => {

        const result: PartialIssue[] = [{
            id: 1,
            assembly: {
                id: 2,
            },
            category: 'a',
            name: 'name',
            subName: 'sub name',
            type: 'a',
            typeName: 'type name',
            typeSubName: 'type sub name',
            status: 'status',
            question: 'question',
            goal: 'goal',
            majorChanges: 'major change',
            changesInLaw: 'changes in law',
            costsAndRevenues: 'costs',
            deliveries: 'deliveries',
            additionalInformation: 'additionalInformation',
            date: '2001-01-01',
            proponents: [{
                id: 3,
                name: 'name',
                avatar: {
                    src: '',
                    templateSrc: '',
                },
                party: {
                    id: 4,
                    name: 'name',
                    color: '123123',
                },
            }],
        }];
        const tree = renderer.create(
            <Router >
                <SearchIssue assembly={1}
                             result={result as Issue[]}
                             isSearching={false}
                             onSearch={() => {}}
                             onClear={() => {}}/>
            </Router>,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
