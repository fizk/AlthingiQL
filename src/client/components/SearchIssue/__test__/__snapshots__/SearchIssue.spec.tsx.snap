// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchIssue /> default props 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <input
    className="input input--md input--search"
    onChange={[Function]}
    onFocus={[Function]}
    type="search"
  />
</div>
`;
